python_prototype/main.py

```diff
@@ -0,0 +1,146 @@
+    return circuit
+
+
+def to_noisy_circuit(circuit: cirq.Circuit, noise_level: float) -> cirq.Circuit:
+    noise_models = [
+        AfterCliffordDepolarizingNoise(noise_level),
+        AfterResetFlipNoise(noise_level),
+        BeforeMeasurementFlipNoise(noise_level),
+    ]
+    for nm in noise_models:
+        circuit = circuit.with_noise(nm)
+    return circuit
+
+
+def generate_stim_circuit(
+    dimension: int, noise_level: float, repetitions: int
+) -> stim.Circuit:
+    template = ScalableQubitSquare(dimension)
+    plaquettes = [
+        ZZPlaquette(PlaquetteOrientation.UP),
+        XXPlaquette(
+            PlaquetteOrientation.LEFT, include_initial_and_final_detectors=False
+        ),
+        ZZZZPlaquette(),
+        XXXXPlaquette(include_initial_and_final_detectors=False),
+        XXPlaquette(
+            PlaquetteOrientation.RIGHT, include_initial_and_final_detectors=False
+        ),
+        ZZPlaquette(PlaquetteOrientation.DOWN),
+    ]
+
+    def make_repeated_layer(circuit: cirq.Circuit) -> cirq.Circuit:
+        any_qubit = next(iter(circuit.all_qubits()), None)
+        assert (
+            any_qubit is not None
+        ), "Could not find any qubit in the given Circuit instance."
+        circuit_to_repeat = (
+            cirq.Circuit([ShiftCoordsGate(0, 0, 1).on(any_qubit)]) + circuit
+        )
+        repeated_circuit_operation = cirq.CircuitOperation(
+            circuit_to_repeat.freeze()
+        ).repeat(repetitions)
+        return cirq.Circuit([repeated_circuit_operation])
+
+    layer_modificators = {1: make_repeated_layer}
+
+    circuit = cirq.Circuit()
+    for layer_index in range(3):
+        layer_circuit = generate_circuit(template, plaquettes, layer_index=layer_index)
+        layer_circuit = normalise_circuit(layer_circuit)
+        circuit += layer_modificators.get(layer_index, lambda circ: circ)(layer_circuit)
+
+    # Compute the qubits that should be measured to get the Z logical operator.
+    plaquette_shape: Shape2D = plaquettes[0].shape
+    assert all(
+        p.shape == plaquette_shape for p in plaquettes
+    ), "All plaquettes should have exactly the same shape for the moment."
+    vertical_center_index: int = (1 + dimension // 2) * (plaquette_shape.y - 1)
+    origin = cirq.GridQubit(vertical_center_index, plaquette_shape.x - 1)
+    circuit.append(
+        [
+            ObservableGate(
+                origin,
+                [
+                    RelativeMeasurement(
+                        cirq.GridQubit(0, i * (plaquette_shape.x - 1)), -1
+                    )
+                    for i in range(dimension + 1)
+                ],
+            ).on(origin)
+        ]
+    )
+    print(circuit)
+    circuit_with_detectors = fill_in_global_record_indices(circuit)
+    noisy_circuit = to_noisy_circuit(circuit_with_detectors, noise_level)
+
+    return cirq_circuit_to_stim_circuit(noisy_circuit)
+
+
+stim_circuit = generate_stim_circuit(2, 0.001, 0)
```

Comment:
Running `python main.py` produces this error repeatedly
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 289, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/local/python/3.10.13/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspaces/tqec/python_prototype/main.py", line 122, in <module>
    collected_surface_code_stats: list[sinter.TaskStats] = sinter.collect(
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection.py", line 380, in collect
    for progress in iter_collect(
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection.py", line 191, in iter_collect
    manager.start_workers(num_workers)
  File "/usr/local/python/3.10.13/lib/python3.10/site-packages/sinter/_collection_work_manager.py", line 85, in start_workers
    w.start()
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/usr/local/python/3.10.13/lib/python3.10/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

Reply:
Yes, I got that too. Seems like a `sinter` error. It works on notebooks, so I ended up using notebooks because solving that was not the top priority.

Reply:
Good point, I saw the option to review the notebooks from this PR. Would you like a review from there as well?

Reply:
I am not the one that configured that. I do not use the ReviewNB that seems to be a third-party service. So I would lean on the "no I do not need you to review on that website" side. If it is simpler for you though, you can do so.

Reply:
Are the notebooks meant for tutorial purposes only? If so, I don't think they need to be rigorously reviewed.

Reply:
Either documentation or as an easy way of testing the results obtained by the code. They can be thought as outside the library, so they can effectively be excluded from review in my opinion.